File: Binary Tree/Binary Search Tree.cpp
#include <bits/stdc++.h>
using namespace std;

class Node
{
    public:
        int node_value;
        Node *left_child;
        Node *right_child;
};

class BinarySearchTree
{
    private:
        Node *root_node;
    public:
        BinarySearchTree()
        {
            root_node = nullptr;
        }
        
        Node* getRootNode()
        {
            return root_node;
        }
        
        void insertNode(int data)
        {
            Node *temp_node;
            if(root_node == nullptr)
            {
                root_node = new Node;
                root_node->node_value = data;
                root_node->left_child = nullptr;
                root_node->right_child = nullptr;
                return;
            }
            else
            {
                Node *sliding_node = root_node;
                Node *not_null_node = root_node;
                while(sliding_node != nullptr)
                {
                    if(sliding_node->node_value > data)
                        {
                            not_null_node = sliding_node;
                            sliding_node = sliding_node->left_child;
                        }
                    else if(sliding_node->node_value < data)
                        {
                            not_null_node = sliding_node;
                            sliding_node = sliding_node->right_child;
                        }
                    else
                        return;
                }
                
                Node *new_node = new Node;
                new_node->node_value = data;
                
                new_node->left_child = nullptr;
                new_node->right_child = nullptr;
                
                if(not_null_node->node_value > data)
                    not_null_node->left_child = new_node;
                else
                    not_null_node->right_child = new_node;
                return;
            }
        }
        
        void inorderTraversal(Node* root_node)
        {
            if(root_node != nullptr)
            {
                inorderTraversal(root_node->left_child);
                cout<<root_node->node_value<<" ";
                inorderTraversal(root_node->right_child);
            }
        }
};

int main()
{
    BinarySearchTree bst;
    bst.insertNode(1);
    bst.insertNode(2);
    bst.insertNode(5);
    bst.insertNode(4);
    bst.insertNode(3);
    bst.insertNode(3);
    
    Node* bst_root_node = bst.getRootNode();
    bst.inorderTraversal(bst_root_node);
    
}

File: Data Structures in Java/Binary Search Tree.java
import java.lang.*;

class Node
{
    public int node_data;
    public Node left_node;
    public Node right_node;
}

class BinarySearchTree
{
    //data member
    private Node root_node = null;
    
    //constructor
    BinarySearchTree()
    {
        root_node = null;
    }
    
    //add a node
    
    public void insert_node(int data)
    {
        if(root_node == null)
        {
            root_node = new Node();
            root_node.node_data = data;
            root_node.left_node = null;
            root_node.right_node = null;
            return;
        }
        else
        {
            Node temp_node = root_node;
            Node trailing_node = new Node();
            trailing_node = null;
            
            Node attached_node = new Node();
            attached_node.node_data = data;
            attached_node.left_node = null;
            attached_node.right_node = null;
            
            while(temp_node != null)
            {
                trailing_node = temp_node;
                if(temp_node.node_data > data)
                    temp_node = temp_node.left_node;
                    
                else if(temp_node.node_data < data)
                    temp_node = temp_node.right_node;
                else
                    return;
                    
            }
            
            if(trailing_node.node_data > data)
                trailing_node.left_node = attached_node;
            else
                trailing_node.right_node = attached_node;
            return;
        }
    }
    
    Node get_root_node()
    {
        return root_node;
    }
    
    public void inorder_traversal(Node root_node)
    {
        if(root_node != null)
        {
            inorder_traversal(root_node.left_node);
            System.out.println(root_node.node_data + " ");
            inorder_traversal(root_node.right_node);
        }
        
    }
    
}

class Main
{
    public static void main(String args[])
    {
        BinarySearchTree b1 = new BinarySearchTree();
        b1.insert_node(10);
        b1.insert_node(5);
        b1.insert_node(20);
        b1.insert_node(50);
        b1.insert_node(70);
        b1.insert_node(40);
        Node b1_root_node = b1.get_root_node();
        b1.inorder_traversal(b1_root_node);
    }
}
